File: spark/src/main/scala/geotrellis/spark/io/index/IndexRanges.scala
package geotrellis.spark.io.index

object IndexRanges {
  /*
   * Will attempt to bin ranges into buckets, each containing at least the average number of elements.
   * Trailing bins may be empty if the count is too high for number of ranges.
   */
  def bin(ranges: Seq[(Long, Long)], count: Int): Seq[Seq[(Long, Long)]] = {
    var stack = ranges.toList

    def len(r: (Long, Long)) = r._2 - r._1 + 1l
    val total = ranges.foldLeft(0l) { (s, r) => s + len(r) }
    val binWidth = total / count + 1

    def splitRange(range: (Long, Long), take: Long): ((Long, Long), (Long, Long)) = {
      assert(len(range) > take)
      (range._1, range._1 + take - 1) -> (range._1 + take, range._2)
    }

    val arr = Array.fill(count)(Nil: List[(Long, Long)])
    var sum = 0l
    var i = 0
    while (stack.nonEmpty) {
      val head = stack.head
      if (len(stack.head) + sum <= binWidth) {
        arr(i) = head :: arr(i)
        sum += len(head)
        stack = stack.tail
      } else {
        val (take, left) = splitRange(head, binWidth - sum)
        stack = left :: stack.tail
        arr(i) = take :: arr(i)
        sum += len(take)
      }

      if (sum >= binWidth) {
        sum = 0l
        i += 1
      }
    }
    arr
  }
}

File: geotrellis-python/geotrellis/spark/io/index/IndexRanges.py
from geotrellis.python.util.utils import fold_left

class IndexRanges(object):

    @staticmethod
    def bin(ranges, count):
        stack = list(ranges)
        def leng(r):
            return r[1] - r[0] + 1
        total = fold_left(ranges, 0, lambda s, r: s + leng(r))
        binWidth = total / count + 1

        def splitRange(_range, take):
            if leng(_range) <= take:
                raise Exception("Assertion failed")
            first = (_range[0], _range[0] + take - 1)
            second = (_range[0] + take, _range[1])
            return (first, second)

        arr = [[]] * count
        summ = 0
        i = 0
        while stack:
            head = stack[0]
            if leng(head) + summ <= binWidth:
                arr[i].insert(0, head)
                summ += leng(head)
                stack = stack[1:]
            else:
                take, left = splitRange(head, binWidth - summ)
                stack[0] = left
                arr[i].insert(0, take)
                summ += leng(take)
            if summ >= binWidth:
                sum = 0
                i += 1
        return arr
            


